src/nf/nf_optimizers.f90

```diff
@@ -187,10 +201,34 @@ pure subroutine minimize_adam(self, param, gradient)
     ! Update parameters.
     param = param &
       - self % learning_rate * m_hat / (sqrt(v_hat) + self % epsilon) &
-      - self % weight_decay_decoupled * param
+      - self % learning_rate * self % weight_decay_decoupled * param
 
     end associate
 
   end subroutine minimize_adam
 
+  impure elemental subroutine init_adagrad(self, num_params)
+    class(adagrad), intent(inout) :: self
+    integer, intent(in) :: num_params
+    if (.not. allocated(self % sum_squared_gradient)) then
+      allocate(self % sum_squared_gradient(num_params))
+      self % sum_squared_gradient = 0
+    end if
+  end subroutine init_adagrad
+
+  pure subroutine minimize_adagrad(self, param, gradient)
+    !! Concrete implementation of an Adagrad optimizer update rule.
+    class(adagrad), intent(inout) :: self
+    real, intent(inout) :: param(:)
+    real, intent(in) :: gradient(:)
+
+    ! Update the sum of squared gradients using the Adagrad rule
+    self % sum_squared_gradient = self % sum_squared_gradient + gradient**2
+
+    ! Update the network parameters based on the new sum of squared gradients
+    param = param - self % learning_rate &
```

Comment:
The learning rate needs to adapt with every time step here, according to https://pytorch.org/docs/stable/generated/torch.optim.Adagrad.html#torch.optim.Adagrad.